### ISSUE TITLE
Scylla does not require ALLOW FILTERING even when the WHERE does not include the partition key

### ISSUE THREAD

**Issue by tzach** [scylladb/scylla#9536]
Scylla version (or git commit hash): 4.5.1

```
create keyspace if not exists ks with replication = {'class': 'SimpleStrategy', 'replication_factor': 1};
create table if not exists ks.tab (p int, c1 int, c2 int, primary key (p, c1, c2));
insert into ks.tab (p, c1, c2) values(0, 0, 0);
insert into ks.tab (p, c1, c2) values(0, 0, 1);
insert into ks.tab (p, c1, c2) values(0, 0, 2);
insert into ks.tab (p, c1, c2) values(0, 1, 0);
insert into ks.tab (p, c1, c2) values(0, 1, 1);
insert into ks.tab (p, c1, c2) values(0, 1, 2);
insert into ks.tab (p, c1, c2) values(0, 2, 0);
insert into ks.tab (p, c1, c2) values(0, 2, 1);
insert into ks.tab (p, c1, c2) values(0, 2, 2);
select c1, c2 from ks.tab where c1 = 1 and c2 = 1;
```

**Comment by slivne** [comment]
this is bad we need to warn the user that he has to place allow filtering - its a full scan with filtering on partial key

**Comment by slivne** [comment]
please  check C* compatibility here - if they allow it - we may need to allow it as well - or useres will have hard time to migrate

**Comment by eliransin** [comment]
> his is bad we need to warn the user that he has to place allow filtering - its a full scan with filtering on partial key

I understand why it is a full scan, what I don't understand is why is it filtering on a partial key? Unless we have a restriction in
the implementation itself (haven't checked) there will be no filtering done in the replica side since the clustering key has been given in full (even though only a prefix should work even though it is partial).

**Comment by eliransin** [comment]
> please check C* compatibility here - if they allow it - we may need to allow it as well - or useres will have hard time to migrate

cassandra 4 indeed gives the filtering error:
```
cqlsh> select c1, c2 from ks.tab where c1 = 1 and c2 = 1;
InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot execute this query as it might involve data filtering and thus may have unpredictable performance. If you want to execute this query despite the performance unpredictability, use ALLOW FILTERING"
```

**Comment by eliransin** [comment]
Indeed Scylla happily accept it. Need to be fixed.

**Comment by nyh** [comment]
@eliransin which Scylla version did you check? If the problem is just on 4.5 but not on master, I think this issue is a **duplicate of #7608**, which was already fixed, we have (several) tests for it, and the fix reached 4.6 but was never backported.
My justification of not backporting this patch was "I don't want to backport this to earlier branches. I don't think it's a great idea that CQL behavior will change in bugfix releases - even if we consider the new behavior more correct.". If @tzach or @eliransin or @slivne feel like this is a real bug and we need to backport it to more releases, I will. But in any case, if this issue is a duplicate of #7608 please close this issue - and ask for the backports on #7608.

BUT, if you see this problem on master as well, then I guess you found a new variant of #7608 which the original patch didn't fix, and we need to write a new test for this new case (let me know if you can really reproduce the problem on master).

**Comment by nyh** [comment]
> I understand why it is a full scan, what I don't understand is why is it filtering on a partial key?

The point is that a "full scan", while it indeed reads a lot of input, also generates a lot of output. This is natural - the amount of work the database needs to do is proportional to the the amount of output. No need to warn the user in such a case of a full scan.

However, Tzach's request `select c1, c2 from ks.tab where c1 = 1 and c2 = 1;`, is different in an important way: It is quite possible that this request will scan the entire table, but return hardly nothing - because you have a billion partitions but none of them actually has the combination `c1 = 1 and c2 = 1` and you can't figure this out without reading the entire table. Here, the amount of work is **not** proportional to the amount of output, and this is exactly the "filtering" that the user needs to approve with the "ALLOW FILTERING" marker.

**Comment by nyh** [comment]
I'm closing this issue as a duplicate of #7608. As I noted above, that issue was already fixed - it just didn't reach version 4.5.1 which Tzach (and I assume Eliran?) tested. We even have a test especially for this example, which I added when Tzach original came up with this scenario:
```python
# Selecting from indexed table using only clustering key should require filterin
g
# Variation of #7608 found in #8991
def test_select_indexed_cluster(cql, table4):
    check_af_mandatory(cql, table4, 'c1 = 1 AND c2 = 1', lambda row: row.c1 == 1 and row.c2 == 1)
```

Whoever wants this fix backported to 4.5, please go to #7608 and say that.

**Comment by nyh** [comment]
@eliransin saw master also fails on this. Need to check why. Reopening.

**Comment by nyh** [comment]
@eliransin actually, like I thought, master is already fixed. But, it doesn't generate an error for the bad request, but rather generates a warning:

**Warnings: This query should use ALLOW FILTERING and will be rejected in future versions.**

Whether this is a warning or an error is controlled by an option "strict_allow_filtering" (see commit 2f28f68e84). It is @avikivity who insisted that this be made a warning, not an error, at least not now. Avi's rationale was that we don't want users who had such (bad...) requests in their application to suddently have some of their requests break after an upgrade. I disagreed with this rationale - I wanted it to be an error because I do consider it an error and don't believe that any reasonable application was written to rely on these requests working. I was also unhappy that our tests check strict_allow_filtering=true while the actual users get strict_allow_filtering=false.

Anyway, what we need to decide now is:
1. Do we need to backport these changes or it's enough that the fix is in 4.6 and master?
2. Should we change the "strict_allow_filtering" option to default to true (error) - not false (warning) - starting in the next version (4.7)? Was it enough that it was a warning for a single version (4.6)?

After deciding on these questions, we should close this issue. It's describes a bug that has already been fixed, and shouldn't remain open as it's written.

**Comment by fee-mendes** [comment]
+1 on default to true, we just stumbled upon this one and got fairly confused.

This is also affecting: `4.5.3-0.20211223.c8f14886d` and `4.4.8-0.20211215.9af2e5ead`, so yes - this needs to be backported @nyh

**Comment by nyh** [comment]
To backport the fix we need to track it in #7608, not here.

I'm closing this issue as a duplicate. If we want to discuss whether this option should be to error by default (as @fee-mendes wanted) instead of warn by default (which was what @avikivity wanted) - then let's create a new issue for that. In the meantime, even a warning would have been better than nothing.
